Docker image digest validation fails on GitHub Actions

From a recent [GitHub Actions build](https://github.com/DataBiosphere/azul/actions/runs/21959203722/job/63438002108#step:5:669):

```
2026-02-12 19:24:02,375    INFO MainThread azul.docker: Resolving 'pycharm' image TagImageRef(registry='docker.io', username='ucscgi', repository=('azul-pycharm',), tag='2025.2.6-72') …
2026-02-12 19:24:02,375    INFO MainThread azul.docker: Resolved 'pycharm' image to DigestImageRef(registry='docker.io', username='ucscgi', repository=('azul-pycharm',), digest='sha256:10d283bdd592e87eb3b8820d5a81aa5670d56bcefd5af3129f822dd5aedb38a9')
2026-02-12 19:24:02,376    INFO MainThread azul.docker: Pulling image DigestImageRef(registry='docker.io', username='ucscgi', repository=('azul-pycharm',), digest='sha256:10d283bdd592e87eb3b8820d5a81aa5670d56bcefd5af3129f822dd5aedb38a9') …
2026-02-12 19:24:23,398    INFO MainThread azul.docker: Pulled image DigestImageRef(registry='docker.io', username='ucscgi', repository=('azul-pycharm',), digest='sha256:10d283bdd592e87eb3b8820d5a81aa5670d56bcefd5af3129f822dd5aedb38a9')
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/runner/work/azul/azul/src/azul/__main__.py", line 34, in <module>
    result = eval(args.expression, dict(__builtins__={}), locals)
  File "<string>", line 1, in <module>
  File "/home/runner/work/azul/azul/src/azul/docker.py", line 720, in resolve_docker_image_for_launch
    assert image.id == parts[str(platform)]['id']
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
…
test (test_tagging.TestTerraformResourceTags.test) ... ok

======================================================================
ERROR: setUpClass (indexer.test_anvil.TestAnvilIndexer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/azul/azul/test/indexer/test_tdr.py", line 194, in setUpClass
    super().setUpClass()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/work/azul/azul/test/azul_test_case.py", line 478, in setUpClass
    super().setUpClass()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/work/azul/azul/test/indexer/__init__.py", line 222, in setUpClass
    super().setUpClass()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/work/azul/azul/test/azul_test_case.py", line 363, in setUpClass
    super().setUpClass()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/work/azul/azul/test/es_test_case.py", line 42, in setUpClass
    image = resolve_docker_image_for_launch('opensearch')
  File "/home/runner/work/azul/azul/src/azul/docker.py", line 720, in resolve_docker_image_for_launch
    assert image.id == parts[str(platform)]['id']
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError

======================================================================
ERROR: setUpClass (indexer.test_anvil.TestAnvilIndexerWithIndexesSetUp)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/azul/azul/test/azul_test_case.py", line 478, in setUpClass
    super().setUpClass()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/work/azul/azul/test/indexer/__init__.py", line 222, in setUpClass
    super().setUpClass()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/work/azul/azul/test/azul_test_case.py", line 363, in setUpClass
    super().setUpClass()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/work/azul/azul/test/es_test_case.py", line 42, in setUpClass
    image = resolve_docker_image_for_launch('opensearch')
  File "/home/runner/work/azul/azul/src/azul/docker.py", line 720, in resolve_docker_image_for_launch
    assert image.id == parts[str(platform)]['id']
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
…

======================================================================
ERROR: setUpClass (service.test_response_anvil.TestAnvilResponse)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/azul/azul/test/service/test_response_anvil.py", line 32, in setUpClass
    super().setUpClass()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/work/azul/azul/test/azul_test_case.py", line 478, in setUpClass
    super().setUpClass()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/work/azul/azul/test/indexer/__init__.py", line 222, in setUpClass
    super().setUpClass()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/work/azul/azul/test/app_test_case.py", line 100, in setUpClass
    super().setUpClass()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/work/azul/azul/test/azul_test_case.py", line 363, in setUpClass
    super().setUpClass()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/work/azul/azul/test/es_test_case.py", line 42, in setUpClass
    image = resolve_docker_image_for_launch('opensearch')
  File "/home/runner/work/azul/azul/src/azul/docker.py", line 720, in resolve_docker_image_for_launch
    assert image.id == parts[str(platform)]['id']
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError

----------------------------------------------------------------------
Ran 289 tests in 129.232s

FAILED (errors=33, skipped=3)
make: *** [Makefile:250: test] Error 1
Error: Process completed with exit code 2.

--- a/src/azul/docker.py
+++ b/src/azul/docker.py
@@ -717,7 +717,15 @@ def resolve_docker_image_for_launch(alias: str) -> str:
         # what specific platform was pulled since we left it to Docker to
         # determine the best match.
         platform = Platform.from_json(image.attrs, config=True).normalize()
-        assert image.id == parts[str(platform)]['id']
+        # When pulling multi-platform images, Docker 29.x with the containerd
+        # image store (default for fresh installations) returns the manifest
+        # digest as image.id instead of the platform-specific config digest.
+        # We accept either to support both storage modes.
+        if ref_to_pull.is_mirrored:
+            image_digest = cast(IndexImageGist, gist)['mirror_digest']
+        else:
+            image_digest = cast(IndexImageGist, gist)['digest']
+        assert image.id in (image_digest, parts[str(platform)]['id']), (image.id, gist)
     # Returning the image ID means that the container will be launched using
     # exactly the image we just pulled and verified.
     return image.id

--- a/src/azul/docker.py
+++ b/src/azul/docker.py
@@ -712,20 +712,23 @@ def resolve_docker_image_for_launch(alias: str) -> str:
         # For single-platform images, this is straight forward.
         assert image.id == cast(ImageGist, gist)['id']
     else:
-        # To determine the expected ID for images that are part of a multi-
-        # platform image aka "manifest list" aka "image index", we need to know
-        # what specific platform was pulled since we left it to Docker to
-        # determine the best match.
-        platform = Platform.from_json(image.attrs, config=True).normalize()
-        # When pulling multi-platform images, Docker 29.x with the containerd
-        # image store (default for fresh installations) returns the manifest
-        # digest as image.id instead of the platform-specific config digest.
-        # We accept either to support both storage modes.
+        # When pulling multi-platform images into the `containerd` image store
+        # (which is the default for fresh installations of Docker 29 or newer),
+        # the ID of the pulled image is set to the digest of the manifest aka
+        # image index. On installations that still use the older `overlay2`
+        # image store, the ID of the pulled image is set to the config digest of
+        # the respective platform-specific image. For now, we'll accept either
+        # to support both image stores. For overlay2, we need to determine what
+        # specific platform was pulled since we left it to Docker to pick the
+        # best match.
+        gist = cast(IndexImageGist, gist)
         if ref_to_pull.is_mirrored:
-            image_digest = cast(IndexImageGist, gist)['mirror_digest']
+            index_digest = gist['mirror_digest']
         else:
-            image_digest = cast(IndexImageGist, gist)['digest']
-        assert image.id in (image_digest, parts[str(platform)]['id']), (image.id, gist)
+            index_digest = gist['digest']
+        platform = Platform.from_json(image.attrs, config=True).normalize()
+        config_digest = parts[str(platform)]['id']
+        assert image.id in (index_digest, config_digest), (image.id, gist)
     # Returning the image ID means that the container will be launched using
     # exactly the image we just pulled and verified.
     return image.id

